### PR TITLE
Add high-level API actions for POA&Ms, SAPs, and SARs

### LINKED ISSUE #56: Add high-level API actions for POA&Ms, SAPs, and SARs
Today, the API specification has support for Catalogs, Profiles, Component Definitions, and SSPs. Some of these have support for "deeper" routes (for example, `/component-definitions/{componentDefinitionId}/parties/{partyId}`) but other document types, we keep it a little more high-level (for now), such as the routes under `/catalogs/`. Adding support for additional document types at a similar level of granularity as a Catalog will at least provide basic support for creating, reading, updating, deleting, and listing these other documents.

### PR DESCRIPTION
**Pull request by nuttercd**
This will close #56

### CODE REVIEW

File: openapi.yaml [hunk 1]
```diff
@@ -1462,7 +1474,7 @@ paths:
         - oscal_auth:
             - write:componentDefinitions
             - read:componentDefinitions
-  /component-definitions/{componentDefinitionId}/components/{componentId}/control-implementations/{componentControlImplementationId}:
+  /component-definitions/{componentDefinitionId}/components/{componentId}/control-implementation/{componentControlImplementationId}:
```
Comment (laurelmay) [inline openapi.yaml]: Why this change?

**Review by laurelmay — DISMISSED** [review]

File: openapi.yaml [hunk 2]
```diff
@@ -2165,93 +2177,752 @@ paths:
...
+                $ref: "#/components/schemas/OSCALPoam"
+        400:
+          description: Invalid ID supplied
+          content: {}
+        404:
+          description: Catalog not found
+          content: {}
+      security:
+        - oscal_auth:
+            - read:catalogs
```
Comment (laurelmay) [inline openapi.yaml]: Let's think carefully about these scopes

File: openapi.yaml [hunk 3]
```diff
@@ -2165,93 +2177,752 @@ paths:
...
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/OSCALPoam"
+        400:
+          description: Invalid ID supplied
+          content: {}
+        404:
+          description: Catalog not found
```
Comment (laurelmay) [inline openapi.yaml]: Not a catalog

File: openapi.yaml [hunk 4]
```diff
@@ -2165,93 +2177,752 @@ paths:
...
+          description: ID of catalog to replace.
+          required: true
+          schema:
+            type: string
+      requestBody:
+        description: Catalog object to be replaced.
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/OSCALCatalog"
```
Comment (laurelmay) [inline openapi.yaml]: I dont think so

File: openapi.yaml [hunk 5]
```diff
@@ -2165,93 +2177,752 @@ paths:
...
+      summary: Deletes an OSCAL Plan of Action and Milestones
+      operationId: deletePoam
+      parameters:
+        - name: api_key
+          in: header
+          schema:
+            type: string
+        - name: poamId
+          in: path
+          description: Catalog ID to delete
```
Comment (laurelmay) [inline openapi.yaml]: Not a catalog

**Review by laurelmay — COMMENTED** [review]
Need need OAuth2 scopes and to fix existing ones.

Good job copying the catalog :D but we need to update some words.

File: openapi.yaml [hunk 6]
```diff
@@ -106,6 +106,18 @@ tags:
     externalDocs:
       description: Find out more
       url: https://pages.nist.gov/OSCAL
+  - name: OSCAL Plan of Action and Milestones
+    externalDocs:
+      description: Find out more
+      url: https://pages.nist.gov/OSCAL/concepts/layer/assessment/poam/
```
Comment (tuckerzp) [inline openapi.yaml:112]: Is there a reason this points to a different url path than above? For example:
```
 - name: OSCAL Catalog
    externalDocs:
      description: Find out more
      url: https://pages.nist.gov/OSCAL/documentation/schema/catalog-layer/catalog/
```

**Review by tuckerzp — CHANGES_REQUESTED** [review]

File: openapi.yaml [hunk 7]
```diff
@@ -2165,93 +2177,752 @@ paths:
...
+          description: Catalog not found
+          content: {}
+      security:
+        - oscal_auth:
+            - write:catalogs
+            - read:catalogs
+  /system-assesment-plan/search:
+    get:
+      tags:
+        - OSCAL System Assessment Results
```
Comment (tuckerzp) [inline openapi.yaml]:
```suggestion
        - OSCAL System Security Plan
```

**Comment by tuckerzp** [comment]
This is a good start! Just make sure to look over your changes in the [`Swagger Editor`](https://editor.swagger.io/?url=https://raw.githubusercontent.com/EasyDynamics/oscal-rest/develop/openapi.yaml) before pushing. That helps a lot to see the places where info doesn't make sense or match up.

File: openapi.yaml (hunk 6 above)
Comment (tuckerzp) [reply openapi.yaml:112]: https://pages.nist.gov/OSCAL/documentation/schema/implementation-layer/component/ goes to https://pages.nist.gov/OSCAL/concepts/layer/implementation/component-definition/

I am asking if you can either make the new `urls` match how we did it in the past, or if we should update the old ones to match what they actually link to.

Comment (nuttercd) [reply openapi.yaml:112]: Oh, I see. I think we should update the old `urls` to match what they actually link to.

File: openapi.yaml (hunk 7 above)
Comment (nuttercd) [reply openapi.yaml]: Why should we change this to Security Plan when its a System Assessment Plan?

**Review by tuckerzp — CHANGES_REQUESTED** [review]
Still looks like some of the tags are incorrectly labeled as `Catalog`

File: openapi.yaml [hunk 8]
```diff
@@ -2252,6 +2889,39 @@ components:
...
+    OSCALPoam:
+      type: object
+      properties:
+        plan-of-action-and-milestones:
+          $ref:
+            "https://raw.githubusercontent.com/EasyDynamics/OSCAL/json-schema-ref-by-path/json/schema/oscal_poam_schema.json#/definitions/assembly_oscal-poam_plan-of-action-and-milestones"
+    OSCALPoamUpdateExample:
+      type: object
+      properties:
+        catalog:
```
Comment (tuckerzp) [inline openapi.yaml]: Should this be catalog?

**Review by tuckerzp — CHANGES_REQUESTED** [review]

File: openapi.yaml (hunk 7 above)
Comment (tuckerzp) [reply openapi.yaml]: I guess it should actually be `OSCAL System Assessment Plans`?

File: openapi.yaml [hunk 9]
```diff
@@ -106,6 +106,18 @@ tags:
     externalDocs:
       description: Find out more
       url: https://pages.nist.gov/OSCAL
+  - name: OSCAL Plan of Action and Milestones
```
Comment (tuckerzp) [inline openapi.yaml:109]: The names have not been plural in the past.

**Review by tuckerzp — CHANGES_REQUESTED** [review]

File: openapi.yaml (hunk 7 above)
Comment (Bronstrom) [reply openapi.yaml]: We're searching for a SAR here, so it makes sense for this to remain as `OSCAL System Assessment Results`.

File: openapi.yaml [hunk 10]
```diff
@@ -2165,6 +2177,631 @@ paths:
...
+        409:
+          description: Conflict with current state
+      security:
+        - oscal_auth:
+            - write:poams
+            - read:poams
+      x-codegen-request-body-name: body
+    put:
+      tags:
+        - OSCAL Plan of Action and Milestones
```
Comment (tuckerzp) [inline openapi.yaml:2296]: These tags are now wrong after the previous change.

File: openapi.yaml [hunk 11]
```diff
@@ -89,23 +89,35 @@ tags:
...
       url: https://pages.nist.gov/OSCAL
+  - name: OSCAL Plan of Action and Milestone
+    externalDocs:
+      description: Find out more
+      url: https://pages.nist.gov/OSCAL/concepts/layer/assessment/poam/
+  - name: OSCAL System Assessment Plan
+    externalDocs:
+      description: Find out more
+      url: https://pages.nist.gov/OSCAL/concepts/layer/assessment/assessment-plan/
+  - name: OSCAL System Assessment Result
```
Comment (tuckerzp) [inline openapi.yaml]: So I know this will take a little bit to change, but should we drop the "System"? 

https://github.com/EasyDynamics/OSCAL/tree/main/json/schema https://pages.nist.gov/OSCAL/reference/latest/complete/json-outline/ 

The files are just named `Assessment Plan` & `Assessment Result`. So I wonder if it is needed? That or we should change the name to `Security Assessment ...`

@kylelaker & @brian-easyd what do you think?

File: openapi.yaml [hunk 12]
```diff
@@ -2165,6 +2177,631 @@ paths:
...
+                type: array
+                items:
+                  $ref: "#/components/schemas/OSCALPoam"
+        400:
+          description: Invalid status value
+          content: {}
+      security:
+        - oscal_auth:
+            - read:poams
+  /system-assessment-plans:
```
Comment (tuckerzp) [inline openapi.yaml]: Again I think the `paths` will need to be singular rather than plural

Comment (tuckerzp) [reply openapi.yaml]: Actually looking at it more, I am wrong here.

File: openapi.yaml (hunk 11 above)
Comment (laurelmay) [reply openapi.yaml]: I'd prefer that align with the NIST terminology here; and drop the first word. "Assessment Plan" and "Assessment Results" to match the documentation and schema.

Comment (aj-stein-nist) [reply openapi.yaml]: And when people casually refer to the meaning of SAP and SAR, it is a Security Assessment Plan and Security Assessment Result(s) in my neck of the woods. Do people actually call it System Assessment Plan in your experience?

Fun fact, and I agree with Dave Waltermire on this: we call them AP and AR and don't mention the S because, in 800-53 land, they can be security assessments, privacy assessments, or other. They aren't strictly always security assessments, and labeling them that way prejudices right from the start.

Also, hi, super excited to see this PR and the discussion here. :-)

Comment (laurelmay) [reply openapi.yaml]: @aj-stein-nist Thanks! I think "System Assessment Plan" was some copying from "System Security Plan" and maybe a misunderstanding of it :D One thing we have noticed is that the [specification _does_ include Security](https://github.com/usnistgov/OSCAL/blob/main/src/metaschema/oscal_assessment-plan_metaschema.xml#L18) but we can open an issue upstream to discuss that if it'd be useful.

I definitely like the idea of not including the "S" to cover the fact it can be for more than one thing!

File: openapi.yaml [hunk 13]
```diff
@@ -2269,3 +2939,9 @@ components:
             read:ssps: read your ssps
             write:parties: modify parties in your account
             read:parties: read your parties
+            write:poams: modify poams in your account
+            read:poams: read your poams
+            write:aps: modify assessment plans in your account
+            read:aps: read your assessment plans parties
+            write:ars: modify assessment results in your account
+            read:ars: read your assessment results in your account
```
Comment (Bronstrom) [inline openapi.yaml]:
```suggestion
            read:aps: read your assessment plans
            write:ars: modify assessment results in your account
            read:ars: read your assessment results
```

File: openapi.yaml [hunk 14]
```diff
@@ -2165,6 +2177,631 @@ paths:
...
+      operationId: replacePoam
+      parameters:
+        - name: poamId
+          in: path
+          description: ID of plan of action and milestone to replace.
+          required: true
+          schema:
+            type: string
+      requestBody:
+        description: Plan of action and milestone object to be replaced.
```
Comment (Bronstrom) [inline openapi.yaml]:
```suggestion
          description: ID of plan of action and milestone to replace
          required: true
          schema:
            type: string
      requestBody:
        description: Plan of action and milestone object to be replaced
```

For consistency purposes, we should delete any periods in the descriptions. There are a couple other places where this occurs

**Review by Bronstrom — CHANGES_REQUESTED** [review]
This is looking really close to completion! I have just a couple minor suggestions.

File: openapi.yaml (hunk 13 above)
Comment (laurelmay) [inline openapi.yaml]: Here and elsewhere, I'd personally prefer to see us use `assessmentPlan`, `assessment-plan`, `assessmentResults`, or `assessment-results` as appropriate. More like Component Definitions than SSPs and POA&Ms.

Thoughts?

Comment (nuttercd) [reply openapi.yaml]: I like the use of `assessmentPlan` as it will match `componentDefinitions`

File: openapi.yaml (hunk 14 above)
Comment (nuttercd) [reply openapi.yaml]: I believe prior descriptions had periods in them. Should we keep or remove periods in descriptions?

File: openapi.yaml [hunk 15]
```diff
@@ -2269,3 +2939,9 @@ components:
             read:ssps: read your ssps
             write:parties: modify parties in your account
             read:parties: read your parties
+            write:poams: modify poams in your account
+            read:poams: read your poams
+            write:assessmentPlan: modify assessment plans in your account
+            read:assessmentPlan: read your assessment plans
```
Comment (laurelmay) [inline openapi.yaml]:
```suggestion
            write:assessmentPlans: modify assessment plans in your account
            read:assessmentPlans: read your assessment plans
```

Fix here and wherever else it's used

File: openapi.yaml (hunk 14 above)
Comment (Bronstrom) [reply openapi.yaml]: Probably best to remove those too, even though they aren't part of this issue.

**Review by Bronstrom — APPROVED** [review]
Nice work @nuttercd!

File: openapi.yaml [hunk 16]
```diff
@@ -2165,6 +2176,631 @@ paths:
       security:
         - oscal_auth:
             - read:parties
+  /plan-of-action-and-milestones:
+    get:
+      tags:
+        - OSCAL Plan of Action and Milestone
+      summary: Returns all OSCAL plan of action and milestones
```
Comment (tuckerzp) [inline openapi.yaml]: I feel like `Plans of action and milestones` sounds better as the plural here.

```suggestion
      summary: Returns all OSCAL plans of action and milestones
```

File: openapi.yaml [hunk 17]
```diff
@@ -2165,6 +2176,631 @@ paths:
...
+                type: array
+                items:
+                  $ref: "#/components/schemas/OSCALPoam"
+      security:
+        - oscal_auth:
+            - read:poams
+    post:
+      tags:
+        - OSCAL Plan of Action and Milestone
+      summary: Adds a new OSCAL plan of action and milestone
```
Comment (tuckerzp) [inline openapi.yaml]: Should `milestone` always be plural? That is how it appears on the [spec ](https://pages.nist.gov/OSCAL/reference/latest/plan-of-action-and-milestones/json-reference/). At the very least, we should be consistent on this. 
```suggestion
      summary: Adds a new OSCAL plan of action and milestones
```

**Review by tuckerzp — CHANGES_REQUESTED** [review]
I think this is almost perfect! Thank you @nuttercd for working on this so far. I am excited to get it in soon.

**Review by tuckerzp — DISMISSED** [review]
Thank you so  much for you work on this @nuttercd! I am very glad to get this in.

**Comment by tuckerzp** [comment]
@kylelaker Looking over everything again, things look good to me. I wanted to give you a chance to look over at things before I dismissed your old review.

File: openapi.yaml [hunk 18]
```diff
@@ -538,9 +549,9 @@ paths:
         - OSCAL Profile
       summary: Removes a catalog from an OSCAL profile imports
       description: |
-        Removes the given catalog from the given profile.
+        Removes the given catalog from the given profile
 
-        This should also result in references under other parts of the schema being removed.
+        This should also result in references under other parts of the schema being removed
```
Comment (tuckerzp) [inline openapi.yaml]: I do not think that _every_ description needs to have periods removed.

**Review by laurelmay — APPROVED** [review]
Thanks for your work on this! I think it's a good start for these types!

File: openapi.yaml (hunk 11 above)
Comment (brian-ruf-ezd) [reply openapi.yaml]: @kylelaker @aj-stein-nist - In FedRAMP parlance (and most Agency implementations of FISMA) they are:
- Security Assessment Plan (SAP); and
- Security Assessment Report (SAR)  [NOTE: Not "result", but "report".]

"System" is not typically used.

That said, Dave and I debated the AP, AR name and he made the very valid point that while these are the correct terms for government use, OSCAL is more than just for the Federal government. He asserted that "Assessment Plan" and "Assessment Result" were more friendly to all frameworks.
